README.md
# I2C-PWM
I2C pulse-width modulation controller module

https://cdn-learn.adafruit.com/downloads/pdf/16-channel-pwm-servo-driver.pdf

CHANGELOG.md
## v1.0.1 (2022-09-18)

### Fix

- Restructure code (#4)

## v1.0.0 (2022-09-17)

### Feat

- Initial release (#3)

## v0.1.0 (2022-09-17)

### Feat

- Initial development
- Initial development
- Initial development
- Initial development

### Fix

- Initial development
- Initial development
- Initial development
- Initial development
- Initial development
- Initial development

## v0.0.0 (2022-08-27)


src/Pca9685Impl.hpp
// Copyright 2022 Kerry Johnson
//
// This file is part of I2C-PWM.
//
// I2C-PWM is free software: you can redistribute it and/or modify it under the
// terms of the GNU General Public License v3 as published by the Free Software
// Foundation.
//
// I2C-PWM is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
// A PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// I2C-PWM.  If not, see https://www.gnu.org/licenses/.
#ifndef PCA9685IMPL_HPP_
#define PCA9685IMPL_HPP_

#include <rclcpp/rclcpp.hpp>
#include <memory>
#include <string>
#include "i2c_pwm/Pca9685.hpp"

namespace i2c_pwm
{

  class Pca9685Impl : public Pca9685
  {
  public:
    Pca9685Impl(const std::string &deviceFile,
                const int address,
                bool autoInitialize = true);
    virtual ~Pca9685Impl();

    void initialize() override;
    void allStop() override;

    void sleepMode(bool value) override;
    void setFrequencyHz(uint16_t value) override;

    uint8_t read(uint8_t reg) const override;
    void write(uint8_t reg, uint8_t data) override;
    void writeChannel(uint8_t channel, uint16_t offValue) override;
    void writeChannel(uint8_t channel, uint16_t onValue, uint16_t offValue)
        override;

  private:
    float pulseWidthToMillis(uint16_t pulseCounts) const;

    const std::string deviceFilePath_;
    const int address_;
    int fd_;
    uint8_t frequencyHz_;
    float frequencyScale_;
    rclcpp::Logger logger_;
  };

} // namespace i2c_pwm

#endif // PCA9685IMPL_HPP_


src/Pca9685Mock.cpp
// Copyright 2022 Kerry Johnson
//
// This file is part of I2C-PWM.
//
// I2C-PWM is free software: you can redistribute it and/or modify it under the
// terms of the GNU General Public License v3 as published by the Free Software
// Foundation.
//
// I2C-PWM is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
// A PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// I2C-PWM.  If not, see https://www.gnu.org/licenses/.
#include "Pca9685Mock.hpp"

namespace i2c_pwm
{

  Pca9685Mock::Pca9685Mock(const std::string & /* deviceFile */,
                           const int /* address */,
                           bool /* autoInitialize */)
      : logger_(rclcpp::get_logger("i2c_pwm.Pca9685Mock"))
  {
  }

  Pca9685Mock::~Pca9685Mock()
  {
  }

  void Pca9685Mock::initialize()
  {
    RCLCPP_INFO(logger_, "Pca9685Mock::initialize()");
  }

  void Pca9685Mock::allStop()
  {
    RCLCPP_INFO(logger_, "Pca9685Mock::allStop()");
  }

  void Pca9685Mock::sleepMode(bool value)
  {
    RCLCPP_INFO(logger_,
                "Pca9685Mock::sleepMode(value=%s)",
                value ? "true" : "false");
  }

  void Pca9685Mock::setFrequencyHz(uint16_t value)
  {
    RCLCPP_INFO(logger_, "Pca9685Mock::setFrequencyHz(value=%d)", value);
  }

  uint8_t Pca9685Mock::read(uint8_t reg) const
  {
    RCLCPP_INFO(logger_, "Pca9685Mock::read(reg=%d)", reg);
    return reg;
  }

  void Pca9685Mock::write(uint8_t reg, uint8_t data)
  {
    RCLCPP_INFO(logger_, "Pca9685Mock::write(reg=%d, data=%d)", reg, data);
  }

  void Pca9685Mock::writeChannel(uint8_t channel, uint16_t offValue)
  {
    RCLCPP_INFO(logger_,
                "Pca9685Mock::writeChannel(channel=%d, offValue=%d)",
                channel,
                offValue);
  }

  void Pca9685Mock::writeChannel(uint8_t channel,
                                 uint16_t onValue,
                                 uint16_t offValue)
  {
    RCLCPP_INFO(logger_,
                "Pca9685Mock::writeChannel(channel=%d, onValue=%d, offValue=%d)",
                channel,
                onValue,
                offValue);
  }

} // namespace i2c_pwm


src/Pca9685Mock.hpp
// Copyright 2022 Kerry Johnson
//
// This file is part of I2C-PWM.
//
// I2C-PWM is free software: you can redistribute it and/or modify it under the
// terms of the GNU General Public License v3 as published by the Free Software
// Foundation.
//
// I2C-PWM is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
// A PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// I2C-PWM.  If not, see https://www.gnu.org/licenses/.
#ifndef PCA9685MOCK_HPP_
#define PCA9685MOCK_HPP_

#include <rclcpp/rclcpp.hpp>
#include <stdint.h>
#include "i2c_pwm/Pca9685.hpp"

namespace i2c_pwm
{

  class Pca9685Mock : public Pca9685
  {
  public:
    Pca9685Mock(const std::string &deviceFile,
                const int address,
                bool autoInitialize);
    virtual ~Pca9685Mock();

    void initialize() override;
    void allStop() override;

    void sleepMode(bool value) override;
    void setFrequencyHz(uint16_t value) override;

    uint8_t read(uint8_t reg) const override;
    void write(uint8_t reg, uint8_t data) override;
    void writeChannel(uint8_t channel, uint16_t offValue) override;
    void writeChannel(uint8_t channel, uint16_t onValue, uint16_t offValue)
        override;

  private:
    rclcpp::Logger logger_;
  };

} // namespace i2c_pwm

#endif /* PCA9685MOCK_HPP_ */


src/i2c_pwm/src/Pca9685.cpp
// Copyright 2022 Kerry Johnson
//
// This file is part of I2C-PWM.
//
// I2C-PWM is free software: you can redistribute it and/or modify it under the
// terms of the GNU General Public License v3 as published by the Free Software
// Foundation.
//
// I2C-PWM is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
// A PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// I2C-PWM.  If not, see https://www.gnu.org/licenses/.
#include "i2c_pwm/Pca9685.hpp"
#include "Pca9685Impl.hpp"
#include "Pca9685Mock.hpp"

namespace
{
  bool mockMode = false;
} // namespace

namespace i2c_pwm
{

  const uint8_t Pca9685::NUM_CHANNELS;
  const uint16_t Pca9685::MIN_VALUE;
  const uint16_t Pca9685::MAX_VALUE;

  std::shared_ptr<Pca9685> Pca9685::create(const std::string &deviceFile,
                                           const int address,
                                           bool autoInitialize)
  {
    if (mockMode)
    {
      return std::shared_ptr<Pca9685>(new Pca9685Mock(deviceFile,
                                                      address,
                                                      autoInitialize));
    }
    else
    {
      return std::shared_ptr<Pca9685>(new Pca9685Impl(deviceFile,
                                                      address,
                                                      autoInitialize));
    }
  }

  void Pca9685::setMockMode(bool value)
  {
    mockMode = value;
  }

}


src/i2c_pwm/src/Pca9685Impl.cpp
// Copyright 2022 Kerry Johnson
//
// This file is part of I2C-PWM.
//
// I2C-PWM is free software: you can redistribute it and/or modify it under the
// terms of the GNU General Public License v3 as published by the Free Software
// Foundation.
//
// I2C-PWM is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
// A PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// I2C-PWM.  If not, see https://www.gnu.org/licenses/.
#include "Pca9685Impl.hpp"
#include <fcntl.h>

extern "C"
{
#include <i2c/smbus.h>
#include <linux/i2c.h>
#include <linux/i2c-dev.h>
}
#include <sys/ioctl.h>
#include <unistd.h>
#include <algorithm>
#include <bitset>
#include <iostream>
#include <iomanip>
#include <stdexcept>
#include <string>

// NOTES:
//
// Right now, this implementation assumes a single I2C board on the
// system (or, more accurately, a single user of the I2C device file).
// Changes would need to be made in order to account for multiple I2C
// users (e.g. convert each Pca9685 to a *Node* or convert this file
// to behave properly in the presence of multiple users... non-blocking
// calls, etc).
//
//    See: https://www.spinics.net/lists/linux-i2c/msg17264.html
//    And: https://stackoverflow.com/a/41187358
//
// For hardware specs, register values, and general board behavior:
//    Ref: https://cdn-shop.adafruit.com/datasheets/PCA9685.pdf

namespace
{

  const float OSC_CLOCK_HZ = 25000000.0; // 25MHz
  const uint16_t MAX_FREQ_MOD_HZ = 1526;
  const uint16_t MIN_FREQ_MOD_HZ = 24;
  const uint16_t DEFAULT_FREQ_HZ = 50;

  const uint8_t REGISTER_MODE1 = 0x00;
  const uint8_t REGISTER_MODE2 = 0x01;
  const uint8_t REGISTER_CHANNEL0_ON_LOW = 0x06;
  const uint8_t REGISTER_CHANNEL0_ON_HIGH = 0x07;
  const uint8_t REGISTER_CHANNEL0_OFF_LOW = 0x08;
  const uint8_t REGISTER_CHANNEL0_OFF_HIGH = 0x09;
  const uint8_t REGISTER_ALL_OFF_HIGH = 0xFD;
  const uint8_t REGISTER_PRE_SCALE = 0xFE;

  const uint8_t RESTART = 0x80;
  const uint8_t CLEAR_RESTART = 0x7F;
  const uint8_t SLEEP = 0x10;
  const uint8_t ALL_FULL_OFF = 0x10;

  // By default, we only want ALLCALL to be enabled (bit 0):
  const uint8_t DEFAULT_MODE1_MASK = 0x01;

  // By default, we only want OUTDRV (totem) to be enabled (bit 2);
  const uint8_t DEFAULT_MODE2_MASK = 0x04;

  // Documentation states "at least" 500us
  const struct timespec OSC_STABILIZATION_DELAY = {0, 600000L};

  const uint8_t LOWER_BYTE_MASK = 0xFF;
  const uint8_t UPPER_BYTE_SHIFT = 8;

} // namespace

namespace i2c_pwm
{

  Pca9685Impl::Pca9685Impl(const std::string &deviceFile,
                           const int address,
                           bool autoInitialize)
      : deviceFilePath_(deviceFile),
        address_(address),
        fd_(0),
        frequencyHz_(0),
        frequencyScale_(1.0f),
        logger_(rclcpp::get_logger("i2c_pwm.Pca9685"))
  {
    if (autoInitialize)
    {
      Pca9685Impl::initialize();
    }
  }

  Pca9685Impl::~Pca9685Impl()
  {
    if (fd_ != 0)
    {
      ::close(fd_);
    }
  }

  void Pca9685Impl::initialize()
  {
    if (fd_ == 0)
    {
      fd_ = ::open(deviceFilePath_.c_str(), O_RDWR); // | O_NONBLOCK);

      if (fd_ < 0)
      {
        std::ostringstream ostr;
        ostr << "Unable to open I2C device file: " << deviceFilePath_;
        throw std::runtime_error(ostr.str());
      }

      if (::ioctl(fd_, I2C_SLAVE, address_) < 0)
      {
        std::ostringstream ostr;
        ostr << "Unable to set I2C address: 0x" << std::hex << address_;
        throw std::runtime_error(ostr.str());
      }

      write(REGISTER_MODE1, DEFAULT_MODE1_MASK);
      write(REGISTER_MODE2, DEFAULT_MODE2_MASK);

      setFrequencyHz(DEFAULT_FREQ_HZ * 2);

      allStop();
    }
  }

  void Pca9685Impl::allStop()
  {
    write(REGISTER_ALL_OFF_HIGH, ALL_FULL_OFF);
  }

  void Pca9685Impl::sleepMode(bool value)
  {
    const uint8_t original_mode = read(REGISTER_MODE1);
    const uint8_t new_mode = original_mode & (value ? SLEEP : ~SLEEP);

    write(REGISTER_MODE1, new_mode);

    nanosleep(&OSC_STABILIZATION_DELAY, NULL);
  }

  void Pca9685Impl::setFrequencyHz(uint16_t value)
  {
    const uint16_t clampedFreqModHz = std::max(MIN_FREQ_MOD_HZ,
                                               std::min(MAX_FREQ_MOD_HZ, value));

    if (clampedFreqModHz != value)
    {
      RCLCPP_ERROR(logger_,
                   "Clamped frequency modulation value %dHz to %dHz instead",
                   value,
                   clampedFreqModHz);
    }

    frequencyHz_ = clampedFreqModHz;

    frequencyScale_ = static_cast<float>(frequencyHz_) / static_cast<float>(DEFAULT_FREQ_HZ);

    //                  ( OSC_CLOCK_HZ )
    // prescale = round( -------------- ) - 1
    //                  (  4096 x rate )
    const uint8_t prescale = std::round(OSC_CLOCK_HZ / (static_cast<float>(MAX_VALUE) * static_cast<float>(clampedFreqModHz))) - 1;

    RCLCPP_DEBUG(logger_,
                 "Setting PWM frequency to %d Hz (prescale: %d)",
                 clampedFreqModHz,
                 prescale);

    const uint8_t oldmode = read(REGISTER_MODE1);

    // PRE_SCALE register can only be updated when SLEEP mode enabled
    RCLCPP_DEBUG(logger_, "MODE1: Sleep");
    write(REGISTER_MODE1, (oldmode & CLEAR_RESTART) | SLEEP);

    RCLCPP_DEBUG(logger_, "PRE_SCALE");
    write(REGISTER_PRE_SCALE, prescale);

    // Restore the original mode, minus SLEEP
    RCLCPP_DEBUG(logger_, "MODE1: Restore");
    write(REGISTER_MODE1, oldmode);

    nanosleep(&OSC_STABILIZATION_DELAY, NULL);

    // restart
    RCLCPP_DEBUG(logger_, "MODE1: Restart...");
    write(REGISTER_MODE1, oldmode | RESTART);
  }

  uint8_t Pca9685Impl::read(uint8_t reg) const
  {
    const int result = i2c_smbus_read_byte_data(fd_, reg);

    if (result < 0)
    {
      RCLCPP_ERROR(logger_,
                   "Unable to read I2C device register 0x%x; errno: %d",
                   reg,
                   result);

      return 0;
    }
    else
    {
      return static_cast<uint8_t>(result);
    }
  }

  void Pca9685Impl::write(uint8_t reg, uint8_t data)
  {
    RCLCPP_DEBUG(logger_,
                 "Writing 0x%02x <-- 0b%s",
                 reg,
                 std::bitset<8>(data).to_string().c_str());

    const int result = i2c_smbus_write_byte_data(fd_, reg, data);

    if (result < 0)
    {
      RCLCPP_ERROR(logger_,
                   "Unable to write I2C device register 0x%x; errno: %d",
                   reg,
                   result);
    }
  }

  void Pca9685Impl::writeChannel(uint8_t channel, uint16_t offValue)
  {
    writeChannel(channel, 0, offValue);
  }

  void Pca9685Impl::writeChannel(uint8_t channel,
                                 uint16_t onValue,
                                 uint16_t offValue)
  {
    if (channel > NUM_CHANNELS)
    {
      std::ostringstream ostr;
      ostr << "Channel out of range: " << channel;
      throw std::runtime_error(ostr.str());
    }

    if (offValue > MAX_VALUE)
    {
      std::ostringstream ostr;
      ostr << "Value out of range: " << offValue;
      throw std::runtime_error(ostr.str());
    }

    if (offValue < onValue)
    {
      std::ostringstream ostr;
      ostr << "Invalid onValue/offValue combination: "
           << onValue
           << ", "
           << offValue;
      throw std::runtime_error(ostr.str());
    }

    const int channelOffset = channel * 4;

    const uint16_t scaledOnValue = onValue * frequencyScale_;
    const uint16_t scaledOffValue = offValue * frequencyScale_;

    RCLCPP_DEBUG(logger_,
                 "Channel %d ON: %d (%0.2fms)",
                 REGISTER_CHANNEL0_ON_LOW + channelOffset,
                 scaledOnValue,
                 pulseWidthToMillis(scaledOnValue));
    RCLCPP_DEBUG(logger_,
                 "Channel %d OFF: %d (%0.2fms)",
                 REGISTER_CHANNEL0_OFF_LOW + channelOffset,
                 scaledOffValue,
                 pulseWidthToMillis(scaledOffValue));

    write(REGISTER_CHANNEL0_ON_LOW + channelOffset,
          scaledOnValue & LOWER_BYTE_MASK);
    write(REGISTER_CHANNEL0_ON_HIGH + channelOffset,
          scaledOnValue >> UPPER_BYTE_SHIFT);
    write(REGISTER_CHANNEL0_OFF_LOW + channelOffset,
          scaledOffValue & LOWER_BYTE_MASK);
    write(REGISTER_CHANNEL0_OFF_HIGH + channelOffset,
          scaledOffValue >> UPPER_BYTE_SHIFT);
  }

  float Pca9685Impl::pulseWidthToMillis(uint16_t pulseCounts) const
  {
    // E.g. 50Hz == 0.002s == 20ms
    const float secondsPerCycle = 1.0f / frequencyHz_;
    const float msPerCycle = secondsPerCycle * 1000.0f;
    const float msPerCount = msPerCycle / static_cast<float>(Pca9685::MAX_VALUE);

    return pulseCounts * msPerCount;
  }

} // namespace i2c_pwm


include/i2c_pwm/Pca9685.hpp
// Copyright 2022 Kerry Johnson
//
// This file is part of I2C-PWM.
//
// I2C-PWM is free software: you can redistribute it and/or modify it under the
// terms of the GNU General Public License v3 as published by the Free Software
// Foundation.
//
// I2C-PWM is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
// A PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// I2C-PWM.  If not, see https://www.gnu.org/licenses/.
#ifndef I2C_PWM__PCA9685_HPP_
#define I2C_PWM__PCA9685_HPP_

#include <memory>
#include <stdint.h>
#include <string>

namespace i2c_pwm
{

  class Pca9685
  {
  public:
    static const uint8_t NUM_CHANNELS = 16;
    static const uint16_t MIN_VALUE = 0;
    static const uint16_t MAX_VALUE = 4096; // 2**12

    static std::shared_ptr<Pca9685> create(const std::string &deviceFile,
                                           const int address,
                                           bool autoInitialize);

    static void setMockMode(bool value);

    virtual ~Pca9685()
    {
    }

    /**
     * (Re-)Initializes the PCA9685 with default settings.
     */
    virtual void initialize() = 0;

    /**
     * Stops all servos.
     */
    virtual void allStop() = 0;

    /**
     * Enables or disables low power (sleep) mode.
     *
     * Low power mode turns off the oscillator.  PWM control is disabled
     * in low-power mode.
     *
     * @param value determines whether sleep mode is enacted
     */
    virtual void sleepMode(bool value) = 0;

    /**
     * Adjust the output modulation frequency from 24Hz to 1526Hz.
     *
     * @param value the frequency (in Hz) [24, 1526]
     */
    virtual void setFrequencyHz(uint16_t value) = 0;

    /**
     * Reads and returns the 8-bit value contained in the given register.
     *
     * @param reg
     * @return
     */
    virtual uint8_t read(uint8_t reg) const = 0;

    /**
     * Writes the given 8-bit value to the specified register.
     *
     * @param reg
     * @param data
     */
    virtual void write(uint8_t reg, uint8_t data) = 0;

    /**
     * Writes (an assumed) onValue of 0 and the given offValue to the registers
     * associated with the specified channel.
     *
     * @param channel the PWM channel (0-based) to write the given data
     * @param offValue the off value for the given channel [0 - 4096]
     */
    virtual void writeChannel(uint8_t channel, uint16_t offValue) = 0;

    /**
     * Writes the given onValue and offValue to the registers
     * associated with the specified channel.
     *
     * @param channel the PWM channel (0-based) to write the given data
     * @param onValue the on value for the given channel [0 - 4096]
     * @param offValue the off value for the given channel [0 - 4096]
     */
    virtual void writeChannel(uint8_t channel,
                              uint16_t onValue,
                              uint16_t offValue) = 0;
  };

} // namespace i2c_pwm

#endif // I2C_PWM__PCA9685_HPP_
